src/mongo/delegates/Updates.class.php

```diff
@@ -659,6 +668,75 @@ protected function applyChangeSet(\Tripod\ChangeSet $cs, $originalCBDs, $context
         }
     }
 
+    private function getAdditionsRemovalsGroupedByNsPredicate($cs,$additions,$removals)
+    {
+        $additionsGroupedByNsPredicate = $this->getChangesGroupedByNsPredicate($cs,$additions);
+        $removalsGroupedByNsPredicate = $this->getChangesGroupedByNsPredicate($cs,$removals);
+
+        $mergedResult = array();
+        foreach($additionsGroupedByNsPredicate as $predicate => $values)
+        {
+            if (!isset($mergedResult[$predicate]))
+            {
+                $mergedResult[$predicate] = array();
+            }
+            $mergedResult[$predicate]["additions"] = $values;
+        }
+        foreach($removalsGroupedByNsPredicate as $predicate => $values)
+        {
+            if (!isset($mergedResult[$predicate]))
+            {
+                $mergedResult[$predicate] = array();
+            }
+            $mergedResult[$predicate]["removals"] = $values;
+        }
+        return $mergedResult;
+    }
+
+    private function getChangesGroupedByNsPredicate(\Tripod\ChangeSet $cs, $changes)
+    {
+        $changesGroupedByNsPredicate = array();
+        foreach ($changes as $c)
+        {
+            $predicate = $cs->get_first_resource($c["value"],$this->labeller->qname_to_uri("rdf:predicate"));
+            $nsPredicate = $this->labeller->uri_to_qname($predicate);
+
+            if(!array_key_exists($nsPredicate, $changesGroupedByNsPredicate))
+            {
+                $changesGroupedByNsPredicate[$nsPredicate] = array();
+            }
+
+            $object = $cs->get_subject_property_values($c["value"],$this->labeller->qname_to_uri("rdf:object"));
+            if (count($object)!=1)
+            {
+                $this->getLogger()->error("Expecting object array with exactly 1 element",$object);
+                throw new \Tripod\Exceptions\Exception("Object of removal malformed");
+            }
+
+            $valueType = (($object[0]['type']=="uri")) ? VALUE_URI : VALUE_LITERAL;
+            $value = ($valueType===VALUE_URI) ? $this->labeller->uri_to_alias($object[0]["value"]) : $object[0]["value"];
+
+            $changesGroupedByNsPredicate[$nsPredicate][] = array($valueType=>$value);
+        }
+        return $changesGroupedByNsPredicate;
+    }
+
+    private function addOperatorToChange(&$changes,$operator,$kvp)
+    {
+        if (!isset($changes[$operator]) || !is_array($changes[$operator]))
+        {
+            $changes[$operator] = array();
+        }
+        foreach($kvp as $key=>$value)
+        {
+            if (isset($changes[$operator][$key]))
+            {
+                $value = array_merge($value,$changes[$operator][$key]);
```

Comment:
I have no idea if this matters, but this will put ` $value ` as the first value of the array.  Would that possibly affect methods, such as get_first_resource/get_first_literal?